Adds missing schemas + data for metrics to snowplow stats ping

escherize opened the pull request:
resolves #48424

Most of the info for this section was in the legacy stats info. The new stuff is calculated in [`->snowplow-metric-info`](https://github.com/metabase/metabase/blob/add-metrics-to-snowplow-ping/src/metabase/analytics/stats.clj#L599).

The specified tags are included with `"metrics"`.

Also, I refactored this to be more repl-friendly. We can call `generate-instance-stats!` from the repl to see its value.

This adds the jsonschema for `"metrics"` section, as well as `"grouped_metrics"`.

escherize committed:
adds data + schema for metrics stats ping

escherize committed:
remove comment

escherize committed:
annotate todos

escherize committed:
fill in the rest of the metrics values

- and add defaults

escherize committed:
fix some definitions + use a single timestamp

escherize committed:
shuffle stuff around to appease the namespace linter

- we aren't reaching into the api API, so the linter is more of a
  formality here.

escherize committed:
Merge branch 'master' into add-metrics-to-snowplow-ping

escherize committed:
update docstring

escherize committed:
fix jsonschema, maybe

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:
```diff
@@ -35,6 +35,32 @@
         ]
       }
     },
+    "metrics": {
+      "description": "Key-value pairs of metrics",
+      "type": "array",
+      "items": {
+        "type": "object",
+        "description": "A single metric attribute",
+        "properties": {
+          "key": {
+            "description": "The key for this attribute",
+            "type": "string",
+            "maxLength": 255
+          },
+          "value": {
+            "description": "The value of this attribute",
+            "type": ["string", "boolean", "integer", "null"],
+            "maxLength": 255,
+            "minimum": 0,
+            "maximum": 2147483647
+          }
+        },
+        "required": [
+          "key",
+          "value"
+        ]
+      }
+    },
```
**nitpick (blocking):** We need to make this a new minor schema version `1-0-1`.

Context: In SnowcatCloud the schemas cannot be edited when they are deployed already. This means if we change it here and start sending new properties (i.e. `metrics`), the event will be rejected. So if we do updates after deployment, they need to be new schema versions that we then deploy again in SnowcatCloud 🙂

escherize replied:
Thank you for pointing this out! 

I will revert this file, and add the `"metrics"` key to an identical json file in `metabase/snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1`. 

I think I also need to bump the value for `::instance_stats` to `"1-0-1"` [here](https://github.com/metabase/metabase/blob/add-metrics-to-snowplow-ping/src/metabase/analytics/snowplow.clj#L35-L41). But I'm not sure: Do you know when a new version of a schema gets added to SnowcatCloud? Are there some docs for this? I'll look around when I get a chance.

Somtom replied:
> I think I also need to bump the value for ::instance_stats to "1-0-1"
Yes I think that, too (but @qnkhuat) might have better knowledge and can confirm.
[Here](https://www.notion.so/metabase/Snowplow-integration-5da1f874beda4153b4fccfa6c1e77caa?pvs=4#723b7505593c4f7fa1e29416a2377ffa) is a bit of docs around the process but I think we need to still improve it a bit (feel free to adjust it based on your learnings). 

I adjusted a small [section](https://www.notion.so/metabase/Snowplow-integration-5da1f874beda4153b4fccfa6c1e77caa?pvs=4#50e90bb17b27409da20a68a48251bf42) to give some indication on when to talk to SnowcatCloud. TL;DR: When you have the PR up and we internally agreed on the schema format, we should already sanity check with them before merging. Sometimes Joao has some last minute concerns. I was pushing for us (data team) having more ownership and being able to update the schemas. That worked for a week and now we don't have the permissions anymore since their system is not very error proof sometimes.

escherize replied:
fixed. Thanks for the links

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:
```diff
@@ -35,6 +35,32 @@
+          },
+          "value": {
+            "description": "The value of this attribute",
+            "type": ["string", "boolean", "integer", "null"],
```
**question:** What would be the cases when the value is `null`? Shouldn't we always get a value for a metric?

escherize replied:
For the `"metrics"` section, I can be sure we don't produce nulls, so I'll remove it.

escherize replied:
Fixed

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-0:
```diff
@@ -35,6 +35,32 @@
+      "items": {
+        "type": "object",
+        "description": "A single metric attribute",
+        "properties": {
```
**question:** 

Since I cannot find it yet: How did we decide about the `tags` property which was listed [here](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#e3d7fdc32d3446df9e486ce8d63da04b)?

escherize replied:
Thanks for the pointer. I don't see any tags in the product doc, but I will add support for them to the schema and stats collecting code.

escherize replied:
Done.

Somtom replied:
Thanks for adding!
Yeah I think there have been many docs in the meanwhile. FYI [This](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#e3d7fdc32d3446df9e486ce8d63da04b) was the one I was referring to.

Somtom reviewed (commented):
Really excited to see our progress on this migration. I left a few questions and one blocking nitpick that should be easy to resolve :)

escherize committed:
review responses

- revert changes to 1-0-0
- add metrics section to new file: 1-0-1
- bump ::instance_stats to "1-0-1"
- add tags into 1-0-1
- make the code return tags (empty for now until we know what to tag things.)

- also fix test that broke from shuffling settings around

escherize committed:
remove a footgun

escherize committed:
Merge branch 'master' into add-metrics-to-snowplow-ping

escherize committed:
Merge branch 'add-metrics-to-snowplow-ping' of github.com:metabase/metabase into add-metrics-to-snowplow-ping

luizarakaki commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,130 @@
+        ]
+      }
+    },
+    "metrics": {
```
Should we add grouped_metrics too?

luizarakaki replied:
So we don't need to change the schema again to add group_metrics

escherize replied:
good call, I'll add it to the schema

escherize replied:
Looking into this some more, we probably want to say that "grouped_metrics" is a required key. So we will have to change this some more. I'll leave in the "grouped_metrics" schema, however once we are sending `grouped_metrics`, I recommend that we alter [snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1](https://github.com/metabase/metabase/pull/48476/files/4e6afe06fd061a5af249fc7745a76e6ef004c456#diff-830f3618a350a9306a891e1713625bab0c5e8d030fdbced4be15544ee9ed4044) line 159 to be 

```diff 
- "required": ["instance_attributes", "features", "metadata", "metrics"]
+ "required": ["instance_attributes", "features", "metadata", "metrics", "grouped_metrics"]
```

Somtom replied:
Yes. Note that if you make this change after the schema has been deployed, then this means we need to create another schema version. 

We should actually double-check with Joao from SnowcatCloud whether adding new required fields works with a minor version upgrade or whether this would be a major one. If major, we should maybe already add it as required fields and send empty data or something.

escherize replied:
see: https://github.com/metabase/metabase/pull/48476#issuecomment-2405802467

I decided that it's better to add the "grouped_metrics" schema when we start sending grouped metrics. We would still have to update the schema to make the grouped metrics section required once implemented. This way we can setup the schema, and make it required at the same time.

escherize committed:
add tags to metrics,

add grouped_metrics to jsonschema 1-0-1

format 1-0-1

escherize committed:
indent

escherize committed:
cljfmt

escherize committed:
Merge branch 'master' into add-metrics-to-snowplow-ping

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,160 @@
+        "type": "object",
+        "description": "A single metric attribute",
+        "properties": {
+          "key": {
+            "description": "The key for this attribute",
+            "type": "string",
+            "maxLength": 255
+          },
+          "value": {
+            "description": "The value of this attribute",
+            "type": ["string", "boolean", "integer"],
+            "maxLength": 255,
+            "minimum": 0,
+            "maximum": 2147483647
+          },
+          "tags": {
+            "description": "Tags for this metric",
+            "type": "array",
+            "items": {
+              "type": "string",
+              "maxLength": 255
+            }
+          }
+        },
+        "required": ["key", "value"]
```
```suggestion
          "key": {
            "description": "The unique name of the metric",
            "type": "string",
            "maxLength": 255
          },
          "value": {
            "description": "The value of the metric",
            "type": ["number"],
            "maxLength": 255,
            "minimum": 0,
            "maximum": 2147483647
          },,
          "tags": {
            "description": "Tags that can be used flagging teams / features the metric belongs to",
            "type": "array",
            "items": {
              "type": "string",
              "maxLength": 255
            }
          }
        },
        "required": ["name", "value"]
```

escherize replied:
This almost all looks fine, except that we need to remember that `cache_num_queries_cached` is actually a string value because it is a binned balue. We need to either change how it's computed or allow strings in here. (example strings are "0", "1-5", etc. from [here](https://github.com/metabase/metabase/blob/add-metrics-to-snowplow-ping/src/metabase/analytics/stats.clj#L54).

So, we could just return the number and not bin it.

escherize replied:
> `"required": ["name", "value"]`

The maps in this array have the keys: "key" "value" and "tags". I don't think we can require a "name" key w/o specifying what it is.

escherize replied:
I've made this into `"required": ["key", "value", "tags"]` since they will always have those keys.

Somtom replied:
>So, we could just return the number and not bin it.

Yes agreed. If we had binned stuff and care about the counts per bin, this would more become a grouped metric. In this case it sounds reasonable to just always return numbers for metrics.

> The maps in this array have the keys: "key" "value" and "tags". I don't think we can require a "name" key w/o specifying what it is.

Ah yeah sorry - good catch. I forgot to change the `key` to `name`. I think in the schema from the docs, we agreed on `name` for metric instead of `key`.  Similar to how we do it for grouped metrics. There you also have a `name`.
Does that make more sense with that context?

escherize replied:
Cool, I will switch it to "name" "value" and "tags", and unbin `cache_num_wueries_cached`.

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,160 @@
+    }
+  },
+  "additionalProperties": false,
+  "required": ["instance_attributes", "features", "metadata"]
```
Should we change this as well to make the new fields required?

escherize replied:
Well, we are adding the _schema_ for grouped metrics here, but not the data for it. so I will make "metrics" required. but we should hold off on "grouped_metrics" (which will be addable w/o updating the schema in a followup).

Somtom replied:
Yep, saw it in the [other discussion](https://github.com/metabase/metabase/pull/48476#discussion_r1795761878) and commented there. 
The approach is fine for me as long as we don't run into issues such that we need to create a major version update for our schema (that always requires more work since it creates a new table in data land). I should have thought about this earlier when we created the first initial version of the schema. Let's double-check with Joao from SnowcatCloud.

escherize replied:
I believe we should not include grouped metrics until they're being created. We will still have to update the schema (to make them required, and change the type from `[object, null]`, to `object`. I have a copy of the grouped_metrics json schema handy and I can paste it into an issue for adding grouped_metrics.

qnkhuat commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,160 @@
+    "vendor": "com.metabase",
+    "name": "instance_stats",
+    "format": "jsonschema",
+    "version": "1-0-0"
```
version here is mismtached, should be `1-0-1` right?

escherize replied:
fixed.

escherize committed:
version should match filename

escherize committed:
Merge branch 'add-metrics-to-snowplow-ping' of github.com:metabase/metabase into add-metrics-to-snowplow-ping

escherize committed:
update instance stats 1-0-1 schema

escherize committed:
require `tags` in grouped_metric + snowcat comment

escherize committed:
Merge branch 'master' into add-metrics-to-snowplow-ping

escherize committed:
fix formatting noise

escherize committed:
update schema to make it valid

escherize committed:
remove grouped_metrics from instance_stats schema

escherize committed:
cljfmt appeasement

escherize commented:
I decided that it's better to add the `"grouped_metrics"` schema when we start sending grouped metrics.

escherize committed:
Merge branch 'master' into add-metrics-to-snowplow-ping

Somtom reviewed (commented):
Thanks man for tackling the comments. I tried to add a bit of clarification for some things. I really love the progress I see. 
There is one small unknown regarding the required fields that we should figure out and a few minor nits but I am already very happy :)

escherize committed:
unbin cache_num_queries_cached value

- alphabetize metrics generation

escherize committed:
we can now guarantee metric values will be ints

escherize commented:
I've tested this with snowplow mini, and it accepts the value 🎉.

Open question:

- According to the snowplow docs, adding a required field (as we are in this PR) is a breaking change and should mean bumping to 2-0-0. However I also read that we generally shouldn't need to do that, so I could use some clarity there.

escherize committed:
jsonschema for instance_uuid, settings, and grouped_metrics

escherize committed:
add analytics_uuid and make it required

escherize committed:
lint + alphabetize instance stats json schema

escherize committed:
update setting key type

- add maxLength to some strings

escherize committed:
lint jsonschema

- Add description to setting.items.tags
- add maxLength, and {min,max}imum to setting.items.value

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,212 @@
+    "vendor": "com.metabase",
+    "name": "instance_stats",
+    "format": "jsonschema",
+    "version": "1-0-1"
```
After [discussion](https://metaboat.slack.com/archives/C04FZTRF457/p1728889142471029) with Joao, it seems a better and safer approach to go with `2-0-0`. Let's change this here and also rename the file.

Apart from that the file looks good to me. I also asked Joao to double-check whether it is valid (in the linked thread above)

escherize replied:
sounds good, done.

Somtom commented on src/metabase/analytics/snowplow.clj:
```diff
@@ -30,15 +30,21 @@
    ::browse_data    "1-0-0"
    ::invite         "1-0-1"
-   ::instance_stats "1-0-0"
+   ::instance_stats "1-0-1"
```
When we rename to `2-0-0` we also should not forget to adjust here

escherize replied:
done

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,212 @@
+  },
+  "additionalProperties": false,
+  "required": [
+    "analytics_uuid",
+    "features",
+    "grouped_metrics",
+    "instance_attributes",
+    "metadata",
+    "metrics",
+    "settings"
```
Based on the discussion with Joao (making this 2-0-0), we should be safe keeping them required. We should just make sure to send empty arrays for the attributes where we don't have the logic implemented yet.

escherize replied:
done.

Somtom reviewed (commented):
Sorry for the back and forth but there were new infos based on the discussion with SnowcatCloud

Somtom commented on snowplow/iglu-client-embedded/schemas/com.metabase/instance_stats/jsonschema/1-0-1:
```diff
@@ -0,0 +1,212 @@
+    "analytics_uuid": {
+      "description": "The UUID for the instance",
+      "type": "string",
+      "maxLength": 36,
+      "minLength": 36
```
https://metaboat.slack.com/archives/C04FZTRF457/p1728979421380599?thread_ts=1728889142.471029&cid=C04FZTRF457

escherize replied:
done

albertoperdomo commented on src/metabase/analytics/stats.clj:
```diff
@@ -574,11 +583,86 @@
+        _ (clear-translation-count!)]
+    {:models                  (t2/count :model/Card :type :model :archived false)
+     :new_embedded_dashboards (t2/count :model/Dashboard
+                                        :enable_embedding true
```
We've since broken this setting up into multiple ones. This seems outdated. No?

Somtom replied:
I think this is according to the [docs](https://www.notion.so/metabase/Migrate-Anonymous-Stats-ping-to-Snowplow-18f155e235d34ba09f45518a866336e2?pvs=4#1414ef173a184569a2278760650971eb) or what specifically are you referring to @albertoperdomo ?

albertoperdomo replied:
There is no setting enable-embedding anymore, we broke this up into three (static, interactive and SDK). See https://github.com/metabase/metabase/pull/48154

albertoperdomo replied:
I will update the doc.

escherize replied:
@albertoperdomo I see how it looks similar to the setting, but that is not the setting, it is the column on the dashboard table. It hasn't changed.

albertoperdomo commented:
Should this not be backported to the 51 release branch so it makes it into 51? I see the no-backport label.

escherize commented:
> Should this not be backported to the 51 release branch so it makes it into 51? I see the no-backport label.

Thanks, I've switched the labels so we will backport it.

escherize committed:
Bump instance stats to 2-0-0

- remove analytics_uuid string length == 36 check
- adds assertion to ensure required fields are set (and it passes)
- adds info for embedding settings

escherize committed:
adds a grouped-metric to stats ping

- adds length info to the schema to pass jsonschema linting

escherize committed:
cljfmt
